### ISSUE TITLE
[MPS] Grouped/depthwise Conv2d gives incorrect results for batch_size > 1 (fp32 only; correct at batch_size=1)

### ISSUE THREAD

**Issue by wahlinski** [pytorch/pytorch#188335]
### 🐛 Describe the bug

A grouped Conv2d with `groups == in_channels` (kernel spanning the full input height, depthwise-style spatial convolution, as used e.g. in EEGNet's spatial conv layer from the braindecode library) produces incorrect results on MPS whenever batch_size > 1, while batch_size == 1 is correct (max_abs_diff == 0 vs CPU).

While small floating-point differences between backends are expected, this magnitude of error is large enough to make MPS-trained models produce meaningfully different results from CPU-trained ones, which would manifest as accuracy/convergence discrepancies, not just numerical noise.

This reproduces with both the nn.Conv2d module and the functional F.conv2d API, with and without bias, and at every batch size tested (2, 3, 4, 8, 16). It is fully deterministic (identical wrong answer across 5 repeated runs with the same inputs) and does not reproduce in float16, only float32.

### Minimal repro
 
```python
import copy
import torch
import torch.nn as nn
 
torch.manual_seed(0)
device = torch.device("mps")
 
conv_cpu = nn.Conv2d(
    in_channels=8, out_channels=16, kernel_size=(256, 1), groups=8, bias=False
).eval()

conv_mps = copy.deepcopy(conv_cpu).to(device).eval()
 
for batch_size in (1, 2):
    x_cpu = torch.rand(batch_size, 8, 256, 1001)
    x_mps = x_cpu.to(device)
    with torch.no_grad():
        out_cpu = conv_cpu(x_cpu)
        out_mps = conv_mps(x_mps).cpu()
    diff = (out_cpu - out_mps).abs().max().item()
    print(f"batch_size={batch_size}: max_abs_diff={diff:.6f}")
```
 
**Output:**
```
batch_size=1: max_abs_diff=0.000000
batch_size=2: max_abs_diff=1.251600
```

### MPS disagrees with itself, not just with CPU
 
MPS run on a batch of 2 disagrees with MPS run on each of those same two samples individually. The same test run entirely on CPU (as a control) shows exact agreement (`0.000000` for both), confirming this likely is MPS-specific.

```python
x = torch.rand(2, 8, 256, 1001).to(device)
x0, x1 = x[0:1], x[1:2]
 
with torch.no_grad():
    out_batch = conv_mps(x)
    out_cpu = conv_cpu(x.cpu())
    out0 = conv_mps(x0)
    out1 = conv_mps(x1)
 
print((out_batch[0:1] - out0).abs().max().item())
print((out_batch[1:2] - out1).abs().max().item())
print((out_cpu[0:1] - out0.cpu()).abs().max().item())
print((out_cpu[1:2] - out1.cpu()).abs().max().item())
```

### Additional findings
 
- **Deterministic**: identical `max_abs_diff` across 5 repeated runs with the same inputs.
- **Independent of `bias`**: reproduces identically with `bias=True` and `bias=False`.
- **Independent of API surface**: reproduces identically via `F.conv2d` and via the `nn.Conv2d` module.
- **fp32-specific**: does **not** reproduce in `float16` (diff ~0.0002, consistent with normal fp16 precision loss, matching at both batch_size=1  and batch_size=2).
- **Per-sample error pattern is batch-size-dependent and not simply "first sample correct, rest wrong"**. 
```
  batch=2  per-sample diffs: [1.5055, 1.3498]
  batch=3  per-sample diffs: [0.9978, 1.3393, 0.0000]
  batch=4  per-sample diffs: [1.5055, 1.3498, 1.2559, 1.2311]
  batch=8  per-sample diffs: [1.4550, 1.3505, 1.2593, 1.2618, 1.3009, 1.2890, 1.3804, 1.2906]
```

### Versions

PyTorch version: 2.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.5.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.14.5 (v3.14.5:5607950ef23, May 10 2026, 07:38:09) [Clang 21.0.0 (clang-2100.0.123.102)] (64-bit runtime)
Python platform: macOS-26.5.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] rotary-embedding-torch==0.8.9
[pip3] torch==2.12.0
[pip3] torch-einops-utils==0.1.2
[pip3] torchaudio==2.11.0
[pip3] torchinfo==1.8.0

cc @ezyang @gchanan @kadeng @msaroufim @malfet @aditvenk @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

**Comment by Isalia20** [comment]
Reproduced with nightly will take a look. Thanks for the report
